# 18 | Production-headroom harvest features

**Kaggriculture • feature research • manual execution**

32 task descriptors and 12 summaries. One crop-HARVEST value expression changes; the scheduler, staffing, watering, clearing and sale rules remain fixed. Conditional feature values are not final cash or an official score.

Run this notebook in **Kaggriculture Manual (verified source)**. Run the two notebooks sequentially, not concurrently. Do not rerun notebooks 05–16.

## Hypothesis and attribution
Harvesting a ready recurring crop may free capacity for the next production event. Add only the conditional released production value. Current watering is held fixed; no future water or guaranteed sale is assumed.

The research plan and feature dictionary define masks, information availability and failure cases. This is an independent ablation, not a combination of the two rounds.

In [1]:
from pathlib import Path
import json, os, signal, subprocess, sys, threading, queue, time
import pandas as pd
import plotly.io as pio
from IPython.display import display
BASE = Path.cwd()
if not (BASE / "run_research.py").is_file():
    BASE = Path.home() / "kaggriculture_two_rounds"
if not (BASE / "run_research.py").is_file():
    raise FileNotFoundError("Open this notebook inside the extracted two-round package.")
sys.path.insert(0, str(BASE))
from visualize import figures, dashboard
pio.renderers.default = "plotly_mimetype"
ROUND = "18"
print("LIVE AWS EXECUTION • round", ROUND)
print("Package:", BASE)
print("Interpreter:", sys.executable)

def run_stage(command, block=None):
    args = [sys.executable, str(BASE / "run_research.py"), command, "--round", ROUND]
    if block is not None:
        args += ["--block", str(block)]
    cap = 150 if command == "screen" else 210 if command == "pair" else 45
    proc = subprocess.Popen(args, cwd=BASE, stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT, text=True, start_new_session=True)
    messages = queue.Queue()
    def reader():
        for line in proc.stdout:
            messages.put(line)
    threading.Thread(target=reader, daemon=True).start()
    started = time.monotonic()
    try:
        while proc.poll() is None or not messages.empty():
            try:
                print(messages.get(timeout=0.2), end="", flush=True)
            except queue.Empty:
                pass
            if time.monotonic() - started > cap:
                raise TimeoutError("Notebook emergency deadline; preserve diagnostics.")
        if proc.returncode:
            raise RuntimeError("Stage failed. Save the notebook and run the bundle command; do not retry unchanged.")
    finally:
        if proc.poll() is None:
            os.killpg(proc.pid, signal.SIGTERM)
            try:
                proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                os.killpg(proc.pid, signal.SIGKILL)
                proc.wait()


LIVE AWS EXECUTION • round 18
Package: /home/sagemaker-user/kaggriculture_two_rounds
Interpreter: /home/sagemaker-user/projects/kaggriculture/.venv/bin/python


## Prior result: preserve it without promoting it
Clear-only lost own cash while increasing margin. All four factorial outcomes were wins in one repeatedly examined block. That is not evidence of leaderboard improvement.

In [2]:
display(pd.read_csv(BASE / "reference/notebook16_endpoints.csv"))
print(json.loads((BASE / "reference/input_review.json").read_text())["verified_bundle_files"], "prior bundle hashes verified during review")

,mode,retirement,early_clear,evidence,coin_margin,coins,local_match_score,opponent_coins,residual_product_units
0,control,0,0,reused_notebook15,758.0,44840.0,1.0,44082.0,0
1,retire,1,0,reused_notebook15,270.0,41510.0,1.0,41240.0,0
2,clear_only,0,1,one_new_responsive_continuation,830.0,42858.0,1.0,42028.0,0
3,renew,1,1,reused_notebook15,1812.0,43451.0,1.0,41639.0,0


50 prior bundle hashes verified during review


## Stage A — tests, exact mechanics and archived activation
120-second worker cap; 348 fixed observations from three development sources. Tests and live source-action checks must pass. A nonactivated screen skips fresh game pairs. No learned selection is performed.

In [3]:
run_stage("screen")
print((BASE / "outputs" / f"round{ROUND}" / "screen/report.json").read_text())

RESULTS_BUNDLE: /tmp/tmp0nd7qnmw/kaggriculture_two_rounds_results.zip
..............................................................
----------------------------------------------------------------------
Ran 62 tests in 6.462s

OK
{"utc": "2026-09-12T23:05:10.806082+00:00", "event": "PREFLIGHT_PASSED", "verified_objects": 10, "proposed_fresh_development_seeds": [2026091217, 2026091218]}
{"utc": "2026-09-12T23:05:10.942183+00:00", "event": "MECHANICS_PASSED", "round": "18", "cases": 288}
{"utc": "2026-09-12T23:05:13.401924+00:00", "event": "HEARTBEAT", "round": "18", "stage": "screen", "elapsed_seconds": 10.0}
{"utc": "2026-09-12T23:05:17.977759+00:00", "event": "SCREEN_SOURCE_SAVED", "round": "18", "completed": 1, "total": 3}
{"utc": "2026-09-12T23:05:23.406396+00:00", "event": "HEARTBEAT", "round": "18", "stage": "screen", "elapsed_seconds": 20.0}
{"utc": "2026-09-12T23:05:24.287295+00:00", "event": "SCREEN_SOURCE_SAVED", "round": "18", "completed": 2, "total": 3}
{"utc": "2026-09-12T

In [4]:
for fig in figures(BASE, ROUND)[:6]:
    fig.show()

## Stage B — preregistered fresh matched blocks
Each cell has a 180-second worker cap and runs at most two games. Controls are reused across rounds only when their lineage and checksums match. A negative result stops later blocks in this round. Missing/stopped blocks are not zero effects. A shared source/engine/runtime failure is an error: stop, save and bundle.

The four balanced blocks cover two development seeds, both seats and two source-distinct opponents; they are not fully crossed and do not establish separate nuisance-factor effects.

### Block 1: Seed 2026091217 • seat 0 • livestock opponent
Completed controls/games are durable checkpoints. Inspect each result before continuing manually, or use Run All for the four registered blocks with automatic negative-result stops.

In [5]:
run_stage("pair", block=1)
print((BASE / "outputs" / f"round{ROUND}" / "pair1/report.json").read_text())

{"utc": "2026-09-12T23:05:32.958557+00:00", "event": "PREFLIGHT_PASSED", "verified_objects": 10, "proposed_fresh_development_seeds": [2026091217, 2026091218]}
{"utc": "2026-09-12T23:05:32.966014+00:00", "event": "PAIR_SKIPPED", "round": "18", "block": 1, "reason": "STOP_NO_ACTION_ACTIVATION"}
{
  "cloud_resources_modified": false,
  "code_sha256": "1449db58e9cd3cfd6fc61850974c0df6099bf01c44c769927b3c0e659c1df8b8",
  "decision": "STOP_NO_ACTION_ACTIVATION",
  "feature_engineering_complete": false,
  "fingerprint": "5a5e199758822dffd0c7f6e23f51f5b4896bd50b4009e1c51a1756180e9a1573",
  "github_updated": false,
  "models_fit": 0,
  "new_games": 0,
  "official_metric_effect_measured": false,
  "official_submission_score": null,
  "output_sha256": {},
  "round": "18",
  "status": "PAIR_SKIPPED",
  "utc": "2026-09-12T23:05:32.961939+00:00",
  "validation_or_holdout_used": false
}



### Block 2: Seed 2026091217 • seat 1 • crop opponent
Completed controls/games are durable checkpoints. Inspect each result before continuing manually, or use Run All for the four registered blocks with automatic negative-result stops.

In [6]:
run_stage("pair", block=2)
print((BASE / "outputs" / f"round{ROUND}" / "pair2/report.json").read_text())

{"utc": "2026-09-12T23:05:34.396278+00:00", "event": "PREFLIGHT_PASSED", "verified_objects": 10, "proposed_fresh_development_seeds": [2026091217, 2026091218]}
{"utc": "2026-09-12T23:05:34.405885+00:00", "event": "PAIR_SKIPPED", "round": "18", "block": 2, "reason": "STOP_NO_ACTION_ACTIVATION"}
{
  "cloud_resources_modified": false,
  "code_sha256": "1449db58e9cd3cfd6fc61850974c0df6099bf01c44c769927b3c0e659c1df8b8",
  "decision": "STOP_NO_ACTION_ACTIVATION",
  "feature_engineering_complete": false,
  "fingerprint": "5a5e199758822dffd0c7f6e23f51f5b4896bd50b4009e1c51a1756180e9a1573",
  "github_updated": false,
  "models_fit": 0,
  "new_games": 0,
  "official_metric_effect_measured": false,
  "official_submission_score": null,
  "output_sha256": {},
  "round": "18",
  "status": "PAIR_SKIPPED",
  "utc": "2026-09-12T23:05:34.401809+00:00",
  "validation_or_holdout_used": false
}



### Block 3: Seed 2026091218 • seat 1 • livestock opponent
Completed controls/games are durable checkpoints. Inspect each result before continuing manually, or use Run All for the four registered blocks with automatic negative-result stops.

In [7]:
run_stage("pair", block=3)
print((BASE / "outputs" / f"round{ROUND}" / "pair3/report.json").read_text())

{"utc": "2026-09-12T23:05:35.839405+00:00", "event": "PREFLIGHT_PASSED", "verified_objects": 10, "proposed_fresh_development_seeds": [2026091217, 2026091218]}
{"utc": "2026-09-12T23:05:35.851982+00:00", "event": "PAIR_SKIPPED", "round": "18", "block": 3, "reason": "STOP_NO_ACTION_ACTIVATION"}
{
  "cloud_resources_modified": false,
  "code_sha256": "1449db58e9cd3cfd6fc61850974c0df6099bf01c44c769927b3c0e659c1df8b8",
  "decision": "STOP_NO_ACTION_ACTIVATION",
  "feature_engineering_complete": false,
  "fingerprint": "5a5e199758822dffd0c7f6e23f51f5b4896bd50b4009e1c51a1756180e9a1573",
  "github_updated": false,
  "models_fit": 0,
  "new_games": 0,
  "official_metric_effect_measured": false,
  "official_submission_score": null,
  "output_sha256": {},
  "round": "18",
  "status": "PAIR_SKIPPED",
  "utc": "2026-09-12T23:05:35.847796+00:00",
  "validation_or_holdout_used": false
}



### Block 4: Seed 2026091218 • seat 0 • crop opponent
Completed controls/games are durable checkpoints. Inspect each result before continuing manually, or use Run All for the four registered blocks with automatic negative-result stops.

In [8]:
run_stage("pair", block=4)
print((BASE / "outputs" / f"round{ROUND}" / "pair4/report.json").read_text())

{"utc": "2026-09-12T23:05:37.281435+00:00", "event": "PREFLIGHT_PASSED", "verified_objects": 10, "proposed_fresh_development_seeds": [2026091217, 2026091218]}
{"utc": "2026-09-12T23:05:37.295706+00:00", "event": "PAIR_SKIPPED", "round": "18", "block": 4, "reason": "STOP_NO_ACTION_ACTIVATION"}
{
  "cloud_resources_modified": false,
  "code_sha256": "1449db58e9cd3cfd6fc61850974c0df6099bf01c44c769927b3c0e659c1df8b8",
  "decision": "STOP_NO_ACTION_ACTIVATION",
  "feature_engineering_complete": false,
  "fingerprint": "5a5e199758822dffd0c7f6e23f51f5b4896bd50b4009e1c51a1756180e9a1573",
  "github_updated": false,
  "models_fit": 0,
  "new_games": 0,
  "official_metric_effect_measured": false,
  "official_submission_score": null,
  "output_sha256": {},
  "round": "18",
  "status": "PAIR_SKIPPED",
  "utc": "2026-09-12T23:05:37.291546+00:00",
  "validation_or_holdout_used": false
}



## Outcome review and stability
Primary: local match outcome. Secondary: coin margin and own cash. The previous conservative cash guard remains separate and unchanged. Read per-block results and their coverage; do not infer independent sample size from turns or task rows.

In [9]:
run_stage("summary")
summary = json.loads((BASE / "outputs" / f"round{ROUND}" / "summary.json").read_text())
display(pd.DataFrame(summary["comparisons"]))
print(json.dumps({k:v for k,v in summary.items() if k != "comparisons"}, indent=2))

{"utc": "2026-09-12T23:05:37.999973+00:00", "event": "ROUND_COMPLETE", "round": "18", "completed_pairs": 0, "planned_pairs": 4}


""


{
  "automatic_promotion": false,
  "completed_pairs": 0,
  "decision": "REVIEW_FRESH_DEVELOPMENT_BLOCKS_NO_PROMOTION",
  "development_seed_groups": 0,
  "limitations": [
    "Only two new development seeds; not an untouched final holdout.",
    "Four balanced, not fully crossed seed/opponent/seat blocks; no isolated factor effects.",
    "Shared controls make comparisons across rounds correlated.",
    "Crop and livestock references are source-distinct, not established leaderboard-strength opponents.",
    "Three-hired-hand callback restriction unchanged.",
    "No fitted weights, feature selection, or significance claim.",
    "Stopped blocks are missing, not zero effects."
  ],
  "official_metric_effect_measured": false,
  "official_submission_score": null,
  "opponents_tested": [],
  "planned_pairs": 4,
  "round": "18",
  "seats_tested": [],
  "status": "ROUND_COMPLETE"
}


In [10]:
all_figures = figures(BASE, ROUND)
for fig in all_figures[6:]:
    fig.show()

In [11]:
html = dashboard(BASE, ROUND, all_figures)
print("Standalone Plotly dashboard:", html)
print(f"ROUND{ROUND}_COMPLETE — no automatic policy promotion")

Standalone Plotly dashboard: /home/sagemaker-user/kaggriculture_two_rounds/outputs/round18/dashboard.html
ROUND18_COMPLETE — no automatic policy promotion


## Save and return
Press **Ctrl+S**. After both rounds, run `python run_research.py bundle` with the previously verified interpreter (exact terminal command in START_HERE.md). Download **kaggriculture_two_rounds_results.zip**. Bundling is also available after an error. Then stop the JupyterLab application, not the space.

No new submission is made. Callback support for more than three hired hands is still unresolved. See FEATURE_RESEARCH.md for external primary references, distinctions between source facts and hypotheses, and the fresh-validation limitations.